### ISSUE TITLE
[BUG] Rebuild function never deletes the previous database backup

### ISSUE THREAD

**Issue by michelebastione** [litedb-org/LiteDB#2450]
**Version**
5.0.18, 5.0.19 

**Describe the bug**
Every time the LiteEngine.Rebuild method is called the database's contents are copied in a brand new backup file named with an incremental numerical suffix. The previous backup file is never deleted and all the copies just keep accumulating on disk.

**Code to Reproduce**
`using var db = new LiteDatabase("Filename=TestDB");
db.Rebuild();`

**Expected behavior**
There should always be only one backup copy of the database on disk while the previous one should be deleted during the Rebuild process, otherwise all these redundant copies will end up occupying precious disk storage, especially if the database size is significant.

**Screenshots/Stacktrace**
![Result of running the example snippet a few times.](https://github.com/mbdavid/LiteDB/assets/50582333/33ecac9e-9060-49c9-92bb-f4ab6f114852)

**Additional context**
I believe this happens because of the way the backupFilename is assigned in the RebuildService.Rebuild method:
`var backupFilename = FileHelper.GetSuffixFile(_settings.Filename, "-backup", true);`
The checkIfExists parameter sohuld not be set to true, as it will trigger the suffix to increment. Instead a check on the existence of the backup and its potential deletion could be performed before creating the new one. This, at least for me, seemed to solve the issue:
```            
FileHelper.Exec(5, () =>
{
    if (File.Exists(backupFilename))
    {
        File.Delete(backupFilename);
    }
    File.Move(_settings.Filename, backupFilename);
});
```

**Comment by mw911** [comment]
I am faced with the same problem. Using 5.0.20

**Comment by Frenchy62620** [comment]
same problem, i prefer to stay in version 5.0.17

**Comment by mw911** [comment]
@michelebastione 
did you only replace the code with
FileHelper.Exec(5, () =>
{
    if (File.Exists(backupFilename))
    {
        File.Delete(backupFilename);
    }
    File.Move(_settings.Filename, backupFilename);
});

or also set this to false?

var backupFilename = FileHelper.GetSuffixFile(_settings.Filename, "-backup", true);

**Comment by michelebastione** [comment]
@mw911 
Yes, you need to set the `checkIfExists` parameter to false or the function will keep adding increasing numerical suffixes to backupName, that's the root of the problem. The second part is to ensure that the backup file doesn't already exists when we rename the original file, or an IOException is thrown. Though you can avoid to check if the file already exists before calling `File.Delete(backupFilename)`, as it doesn't raise any exception when you do so.

**Comment by sugu0620** [comment]
is this issue resolved in latest build 5.0.21

**Comment by nikubesliu** [comment]
I'm still getting it in 5.0.21.

**Comment by hackerzhuli** [comment]
any progress?

**Comment by MarcusTastler** [comment]
Is a correction in sight? We need a fix...

Workaround/Hotfix:
 ```
string[] backupFilePaths = Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, "MyDbName-backup*.db");
foreach (string filePath in backupFilePaths)
{
  logger.LogWarning($"Remove backup file '{Path.GetFileName(filePath)}'");
  File.Delete(filePath);
}
```

**Comment by JKamsker** [comment]
A fork of LiteDB has worked on this:

- [viveleltsi/LiteDB](https://github.com/viveleltsi/LiteDB), branch [`bugfix/invalide-file-corruption`](https://github.com/viveleltsi/LiteDB/tree/bugfix/invalide-file-corruption): 2 commits not in upstream ([compare with `dev`](https://github.com/litedb-org/LiteDB/compare/dev...viveleltsi:LiteDB:bugfix/invalide-file-corruption))
- [viveleltsi/LiteDB](https://github.com/viveleltsi/LiteDB), branch [`improvement/option-no-backup-migration`](https://github.com/viveleltsi/LiteDB/tree/improvement/option-no-backup-migration): 1 commit not in upstream ([compare with `dev`](https://github.com/litedb-org/LiteDB/compare/dev...viveleltsi:LiteDB:improvement/option-no-backup-migration))

Upstream pull requests from this fork: #2432 (open).

**What it does:** Don't leave a `<name>-backup.db` behind after an automatic upgrade: `EngineSettings.CreateBackupOnUpgrade` and `RebuildOptions.CreateBackup` delete the backup once the rebuild has succeeded.

**Would porting it pay off?** Partially: the need is real, but the fork's code can't be taken as it is.

- It fits dev almost verbatim, since `RebuildService.Rebuild` has the same shape (lines 90-111).
- Before porting, also delete `backupLogFilename`, not just the data backup.
- Apply the option to `Rebuild()` and auto-rebuild, not only to `Upgrade`.
- Combine it with the fixed backup name from #2450 (`GetSuffixFile(..., false)` plus delete-before-move), so that `-backup-N` copies stop piling up.
- Name the option `KeepBackup` (default `true`), as suggested there.
- Fold it into the #2450 backup fix; PR #2432 has the same idea.

Credit: Dan Bourquin (Co-authored-by: Dan Bourquin).

**Comment by JKamsker** [comment]
Triage on `dev` @ faad53821: **still reproduces**.

Related draft PR: #2923 (file ownership, rebuild isolation and WAL identity). It is not merged.
